lib/watch-stream.js

```diff
@@ -0,0 +1,328 @@
+  }
+
+  // if consumed as an async iterator, use that to slow down further...
+  async * [Symbol.asyncIterator] () {
+    const ite = Readable.prototype[Symbol.asyncIterator].call(this)
+
+    for await (const data of ite) {
+      this._yielding = new Promise((resolve) => { this._resolveYield = resolve })
+      yield data
+      this._resolveYield()
+      this._resolveYield = null
+    }
+  }
+
+  // triggered by bee
+  _ontruncate () {
+    // TODO: change the truncated stuff to an ancestor count thing
+    this._truncated = true
+    this._onappend()
+  }
+
+  // triggered by bee
+  _onappend () {
+    // TODO: this is a light hack / fix for non-sparse session reporting .length's inside batches
+    // the better solution is propably just to change non-sparse sessions to not report a fake length
+    if (!this.core.isAutobase && (!this.core.core || this.core.core.tree.length !== this.core.length)) return
+
+    const resolve = this._resolveOnChange
+    this._resolveOnChange = null
+    this._latestChanges++
+    if (resolve) resolve()
+  }
+
+  async _setup () {
+    // overwrite me
+  }
+
+  async _teardown () {
+    // overwrite me
+  }
+
+  async _check () {
+    // overwrite me
+    return null
+  }
+
+  async _open (cb) {
+    try {
+      await this.opened
+    } catch (err) {
+      return cb(err)
+    }
+    cb(null)
+  }
+
+  async _read (cb) {
+    let data = null
+
+    try {
+      while (!this.destroying && data === null) {
+        await this._waitForChanges()
+        if (this.destroying) break
+        await this._yielding
+        if (this.destroying) break
+        this._changes = this._latestChanges
+        const truncated = this._truncated
+        this._truncated = false
+        data = await this._check(truncated)
+      }
+    } catch (err) {
+      return cb(err)
+    }
+
+    if (data !== null) this.push(data)
+    cb(null)
+  }
+
+  _predestroy () {
+    this._teardownPromise = this._teardown()
+    this._teardownPromise.catch(safetyCatch)
+
+    // trigger any pending io
+    if (this._resolveYield) this._resolveYield()
+    this._onappend()
+  }
+
+  async _destroy (cb) {
+    const top = this.bee._watchers.pop()
+
+    if (top !== this) {
+      top._index = this._index
+      this.bee._watchers[top._index] = top
+    }
+
+    if (!this._teardownPromise) this._teardownPromise = this._teardown()
+
+    try {
+      await this._teardownPromise
+    } catch (err) {
+      return cb(err)
+    }
+
+    cb(null)
+  }
+
+  async _waitForChanges () {
+    if (this.destroying) return
+
+    if (this._eager === false && this._changes === this._latestChanges) {
+      await new Promise(resolve => {
+        this._resolveOnChange = resolve
+      })
+    }
+
+    this._eager = false
+  }
+}
+
+class RangeWatchStream extends WatchStream {
+  constructor (bee, range, opts = {}) {
+    super(bee, opts)
+
+    this.range = range
+    this.current = null
+    this.previous = null
+    this.map = opts.map || null
```

Comment:
map and unmap are already set in the superclass. Would either handle completely there, or completely in the subclasses.

`KeyWatchStream` only seems to have `map` and no `unmap`, and no concept of `_mapped` either, so probably makessense to keep in subclass

Reply:
The opened logic could live fully in the superclass btw, and the subclasses just need to overwrite `_setup`--now it lives in both (sueprclass still has `this.opened`)